evalml/pipelines/utils.py

```diff
@@ -1348,3 +1349,109 @@ def rows_of_interest(
+            If set to false, will discard the time index information entirely.
+
+    Returns:
+        pd.DataFrame, pd.DataFrame: The unstacked X and y data.
+    """
+    # Save the target name so we can re-separate it
+    target_name = y.name
+    if target_name is None:
+        target_name = "target"
+        y.name = "target"
+
+    # Combine X and y to make it easier to unstack
+    full_dataset = pd.concat([X, y.set_axis(X.index)], axis=1)
+
+    # Get the total number of series, with their names
+    series_id_unique = full_dataset[series_id].unique()
+
+    # Perform the unstacking
+    X_unstacked_cols = []
+    y_unstacked_cols = []
+    for i in series_id_unique:
+        single_series = full_dataset[full_dataset[series_id] == i]
+
+        # Save the time_index for alignment
+        new_time_index = single_series[time_index]
+        for column_name in full_dataset.columns:
+            if column_name in [time_index, series_id]:
+                continue
+
+            new_column = single_series[column_name]
+            new_column.index = new_time_index
+            new_column.name = f"{column_name}_{i}"
+
+            if column_name == target_name:
+                y_unstacked_cols.append(new_column)
+            else:
+                X_unstacked_cols.append(new_column)
+
+    # Concatenate all the single series to reform dataframes
+    X_unstacked = pd.concat(X_unstacked_cols, axis=1)
+    y_unstacked = pd.concat(y_unstacked_cols, axis=1)
+
+    # Reset the axis if need be
+    if not keep_time_in_index:
+        X_unstacked.reset_index(drop=True, inplace=True)
+        y_unstacked.reset_index(drop=True, inplace=True)
+
+    return X_unstacked, y_unstacked
+
+
+def stack_data(data, include_series_id=False, series_id_name=None):
+    """Stacks the given DataFrame back into a single Series, or a DataFrame if include_series_id is True.
+
+    Should only be used for data that is expected to be a single series. To stack multiple unstacked columns,
+    call this function multiple times on the desired subsets.
+
+    Args:
+        data (pd.DataFrame): The data to stack.
+        include_series_id (bool): Whether or not to extract the series id and include it in a separate columns
+        series_id_name (str): If include_series_id is True, the series_id name to set for the column. The column
+            will be named 'series_id' if this parameter is None.
+
+    Returns:
+        pd.Series or pd.DataFrame: The data in stacked series form.
+    """
+    if data is None or isinstance(data, pd.Series):
+        return data
+
+    stacked_series = data.stack(0)
+
+    # Extract the original column name
+    series_id_with_name = stacked_series.index.droplevel()
+    stacked_series.name = "_".join(series_id_with_name[0].split("_")[:-1])
+
+    # If the index is the time index, keep it
+    if not data.index.is_numeric():
+        new_time_index = data.index.unique().repeat(len(data.columns))
```

Comment:
I was worried that `unique()` doesn't preserve order...but it does.  ------* the more you know